### PR TITLE
U2f support

### PR DESCRIPTION
**Pull request by NickeZ**
TODO
- [x] Store counter in secure element instead of in flash
- [x] clang format

### CODE REVIEW

File: test/unit-test/test_u2f_standard.c [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    abort();
+}
+
+int main(void)
+{
+    // srand((unsigned int) time(NULL));
+
+    //// Test the C code API
+    // U2Fob_testLiveDevice(0);
+    //__stack_chk_guard = random_uint32(0);
```
Comment (benma) [inline test/unit-test/test_u2f_standard.c]: either set the guard or remove the stack_* stuff above.

is the stack protector even enabled in unit tests?

File: src/u2f.h [hunk 2]
```diff
@@ -15,17 +15,21 @@
 #ifndef _U2F_DEVICE_H_
 #define _U2F_DEVICE_H_
 
-#include "usb/u2f/u2f.h"
-#include "usb/usb.h"
 #include <stdbool.h>
```
Comment (benma) [inline src/u2f.h]: remove all includes here, none are used by the header

File: src/u2f.h [hunk 3]
```diff
@@ -15,17 +15,21 @@
...
-#define U2F_HIJACK_ORIGIN_TOTAL 2
+#include <usb/u2f/u2f.h>
+#include <usb/usb.h>
+#include <usb/usb_frame.h>
 
-extern const uint8_t U2F_HIJACK_CODE[U2F_HIJACK_ORIGIN_TOTAL][U2F_NONCE_LENGTH];
+#define U2F_HIJACK_ORIGIN_TOTAL 4
 
-void u2f_device_timeout(void);
+extern const uint8_t U2F_HIJACK_CODE[U2F_HIJACK_ORIGIN_TOTAL][U2F_APPID_SIZE];
```
Comment (benma) [inline src/u2f.h]: unused?

File: src/u2f.c [hunk 4]
```diff
@@ -11,477 +11,606 @@
...
-//    }
-//}
+#define U2F_KEYHANDLE_LEN (U2F_NONCE_LENGTH + SHA256_LEN)
+
+#if (U2F_EC_KEY_SIZE != SHA256_LEN) || (U2F_EC_KEY_SIZE != U2F_NONCE_LENGTH)
+#error "Incorrect macro values for u2f"
+#endif
+
+static uint32_t _cid = 0;
+volatile bool _state_continue = false;
```
Comment (benma) [inline src/u2f.c]: `static` here and next two lines

File: src/u2f.c [hunk 5]
```diff
@@ -11,477 +11,606 @@
...
+    case U2F_VERSION:
+        _version(apdu, out_packet);
+        break;
+    default:
+        _error(U2F_SW_INS_NOT_SUPPORTED, out_packet);
+        return;
+    }
+}
+
+// TODO - currently not called - see v1 code for example usage
```
Comment (benma) [inline src/u2f.c]: delete? seems like you solved timeouts differently

**Comment by NickeZ** [comment]
I changed the logic regarding the "this keyhandle does not belong to this key" case. Now we print an error message when the authentication fails, and then we silently ignore the bogus registrations. I hope that is the most intuitive behavior.

File: CMakeLists.txt [hunk 6]
```diff
@@ -265,6 +265,8 @@ set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror -W -Wall -Wextra -Wpedantic -Wmissin
 
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -ffunction-sections -fdata-sections -fstack-protector-all")
 
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_XOPEN_SOURCE=600")
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply CMakeLists.txt]: done

File: src/keystore.h [hunk 7]
```diff
@@ -178,4 +178,12 @@ USE_RESULT bool keystore_secp256k1_sign(
     uint8_t* sig_compact_out,
     int* recid_out);
 
+/**
+ * Get the seed to be used for u2f
+ * @param seed_out Buffer for seed, must be SHA256_LEN
+ * @param seed_length Length of data written to seed_out
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/keystore.h]: done

File: src/keystore.h [hunk 8]
```diff
@@ -178,4 +178,12 @@ USE_RESULT bool keystore_secp256k1_sign(
     uint8_t* sig_compact_out,
     int* recid_out);
 
+/**
+ * Get the seed to be used for u2f
+ * @param seed_out Buffer for seed, must be SHA256_LEN
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/keystore.h]: done

File: src/securechip/securechip.h [hunk 9]
```diff
@@ -100,4 +102,36 @@ bool securechip_monotonic_increments_remaining(uint32_t* remaining_out);
  */
 bool securechip_random(uint8_t* rand_out);
 
+/**
+ * Generates the matching public key to the provided private key. Will put private key in unsafe
+ * ECC slot.
+ * @param[in] priv_key Private key
+ * @param[out] pub_key Public key
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/securechip/securechip.h]: done

File: src/securechip/securechip.h [hunk 10]
```diff
@@ -100,4 +102,36 @@ bool securechip_monotonic_increments_remaining(uint32_t* remaining_out);
...
+ * @return True if success
+ */
+bool securechip_ecc_generate_public_key(uint8_t* priv_key, uint8_t* pub_key);
+
+/**
+ * Sign hash with private key. Will put private key in unsafe ECC slot.
+ * @param[in] priv_key Private key to use for signing (32 bytes)
+ * @param[in] hash Hash to sign (32 bytes)
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/securechip/securechip.h]: done

File: src/securechip/securechip.c [hunk 11]
```diff
@@ -518,3 +535,134 @@ bool securechip_random(uint8_t* rand_out)
 {
     return atcab_random(rand_out) == ATCA_SUCCESS;
 }
+
+bool securechip_ecc_generate_public_key(uint8_t* priv_key, uint8_t* pub_key)
+{
+    ATCA_STATUS result = _authorize_key();
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/securechip/securechip.c]: done

File: src/securechip/securechip.c [hunk 12]
```diff
@@ -518,3 +535,134 @@ bool securechip_random(uint8_t* rand_out)
...
+    _interface_functions->get_encryption_key(encryption_key);
+
+    result = atcab_priv_write(
+        SECURECHIP_SLOT_ECC_UNSAFE_SIGN,
+        atca_priv_key,
+        SECURECHIP_SLOT_ENCRYPTION_KEY,
+        encryption_key);
+    if (result != ATCA_SUCCESS) {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/securechip/securechip.c]: done

File: src/securechip/securechip.c [hunk 13]
```diff
@@ -518,3 +535,134 @@ bool securechip_random(uint8_t* rand_out)
...
+
+    result = atcab_read_enc(slot, block, bytes, encryption_key, SECURECHIP_SLOT_ENCRYPTION_KEY);
+
+    if (result != ATCA_SUCCESS) {
+        return false;
+    }
+    return true;
+}
+
+static bool _write_data_slot_block(uint8_t* bytes, uint16_t slot, uint8_t block)
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/securechip/securechip.c]: done

File: src/u2f.c [hunk 14]
```diff
@@ -11,477 +11,586 @@
...
-//#include "commander.h"
-//
-//#include "u2f/u2f.h"
-//#include "u2f/u2f_keys.h"
-//
-//#define APDU_LEN(A)              (uint32_t)(((A).lc1 << 16) + ((A).lc2 << 8) + ((A).lc3))
+#include <keystore.h>
+#include <memory.h>
+#include <random.h>
+#include <screen.h>
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/u2f.c]: done

File: src/usb/u2f/u2f_hid.h [hunk 15]
```diff
@@ -1,30 +1,44 @@
 // Copyright 2014 Google Inc. All rights reserved.
-// Copyright 2017-2019 Shift Cryptosecurity Ag
+// Copyright 2017 Douglas J. Bakkum, Shift Devices AG
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/usb/u2f/u2f_hid.h]: done

File: src/usb/usb_frame.c [hunk 16]
```diff
@@ -32,21 +34,37 @@ static void _read_data(State* state, const uint8_t* data, size_t length)
     state->buf_ptr += length;
 }
 
+#include "screen.h"
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/usb/usb_frame.c]: done

File: src/usb/usb_frame.c [hunk 17]
```diff
@@ -32,21 +34,37 @@ static void _read_data(State* state, const uint8_t* data, size_t length)
...
+
 /**
  * Initializes the USB frame packet.
  */
 static int32_t _cmd_init(const USB_FRAME* frame, State* state)
 {
-    if (state->initialized) {
-        return ERR_UNEXPECTED_CMD_INIT;
+    // screen_sprintf_debug(1000, "%d %x %x %x %x", state->initialized, state->cmd, frame->type,
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/usb/usb_frame.c]: done

File: test/unit-test/test_u2f_standard.c (hunk 1 above)
Comment (NickeZ) [reply test/unit-test/test_u2f_standard.c]: not used I think.

File: src/u2f.h (hunk 2 above)
Comment (NickeZ) [reply src/u2f.h]: done

File: src/u2f.h (hunk 3 above)
Comment (NickeZ) [reply src/u2f.h]: Yeah, I left a TODO comment.

File: src/u2f.c (hunk 4 above)
Comment (NickeZ) [reply src/u2f.c]: removed lines instead, dead code.

File: src/u2f.c (hunk 5 above)
Comment (NickeZ) [reply src/u2f.c]: done

File: src/securechip/securechip.h [hunk 18]
```diff
@@ -100,4 +102,36 @@ bool securechip_monotonic_increments_remaining(uint32_t* remaining_out);
  */
 bool securechip_random(uint8_t* rand_out);
 
+/**
+ * Generates the matching public key to the provided private key. Will put private key in unsafe
+ * ECC slot.
+ * @param[in] priv_key Private key
```
[the remark this answers is not in the document]
Comment (NickeZ) [reply src/securechip/securechip.h]: done

**Comment by NickeZ** [comment]
@benma Thanks for the review! This is ready for a new round of reviews.

File: src/u2f.h (hunk 3 above)
Comment (NickeZ) [reply src/u2f.h]: ok! done

**Pull request merged**